.github/workflows/everything.yml

```diff
@@ -131,6 +135,16 @@ jobs:
         with:
           ref: ${{ github.event.pull_request.head.sha }}
           path: source
+      - name: Restore cache
+        uses: actions/cache@v3
+        with:
+          path: .ccache
+          key: ccache-${{ matrix.os }}-${{ matrix.compiler }}-${{ matrix.parallel }}-${{ github.ref }}-${{ github.sha }}
```

Comment:
if we add the sha/ref then new commits will not be able to find the cache. MPI/Serial builds share a lot of cacheable units. I would as far as `ccache-${{ matrix.os }}-${{ matrix.compiler }}`

Reply:
The sha/ref is so that each build can write to a unique cache key.  According to [this](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#using-the-cache-action) section of the github docs, new commits *will* be able to find the cache thanks to the `restore-keys`.

If I remove the mpi/serial bit from the key, I think it could reduce cache hit rate for ompi builds, since sometimes a serial build will write the cache without any of the mpi object files.

Maybe let's discuss key selection strategy when we sync up this afternoon?

Reply:
> If I remove the mpi/serial bit from the key, I think it could reduce cache hit rate for ompi builds, since sometimes a serial build will write the cache without any of the mpi object files.

ccache will deal with this, in fact with ccache we want most builds in the same platform to share the cache directory (key). Ccache generates an unique file for each .o file where its filename is a hash of its path and its compiler invokation with its arguments, and the source file has. There is no chance of overrwriting objects files from one build to the other.

Reply:
> Maybe let's discuss key selection strategy when we sync up this afternoon?

Sure

Reply:
I just found this as I was reading through the docs again:  Since cache is read-only (meaning we can't change existing cache entries, only create new ones), they provided a tip on how to update cache, [here](https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache).

That could be why the pattern of using a unique key for each build seems so frequently used.  

Looking at our cache [usage](https://github.com/ornladios/ADIOS2/actions/caches) now, with only this PR contributing, we're already over the 10GB limit.  Adding up the different cache sizes that were restored in one workflow run, it amounted to around 3GB.  So it seems to me we could easily end up with PR caches pushing out the more important ones from the default branch.

So I think I'm already leaning toward saving caches only on workflows that run on `master`, but restore caches on all branches.  Using the tip linked above, we'd "update" the cache on every master run, then PRs can just use it and hopefully get fairly high hit rates.

But let me know what you think.

Reply:
> I just found this as I was reading through the docs again: Since cache is read-only (meaning we can't change existing cache entries, only create new ones), they provided a tip on how to update cache, [here](https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache).

I see, then no need to swim against the stream here.

> So I think I'm already leaning toward saving caches only on workflows that run on master, but restore caches on all branches. Using the tip linked above, we'd "update" the cache on every master run, then PRs can just use it and hopefully get fairly high hit rates.

That is a sound plan.